copy/copy.go

```diff
@@ -0,0 +1,165 @@
+package copy
+
+import (
+	"crypto/sha256"
+	"crypto/subtle"
+	"encoding/hex"
+	"fmt"
+	"hash"
+	"io"
+	"strings"
+
+	"github.com/containers/image/image"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/transports"
+	"github.com/containers/image/types"
+)
+
+// supportedDigests lists the supported blob digest types.
+var supportedDigests = map[string]func() hash.Hash{
+	"sha256": sha256.New,
+}
+
+type digestingReader struct {
+	source           io.Reader
+	digest           hash.Hash
+	expectedDigest   []byte
+	failureIndicator *bool
```

Comment:
That reads wrong to me, I read verbs as method names.

The lazy way to deal with this is to say that this has not been modified from skopeo’s `copy` but that’s a poor excuse :)

Actually, this does not need to be a pointer at all, simply a bool `validationFailed`.

Still, for consistency I’d prefer moving the code as close to unmodified as possible, and I promise to fix that separately.